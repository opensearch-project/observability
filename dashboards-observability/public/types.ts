/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import { DashboardStart } from '../../../src/plugins/dashboard/public';
import { NavigationPublicPluginStart } from '../../../src/plugins/navigation/public';

export interface AppPluginStartDependencies {
  navigation: NavigationPublicPluginStart;
  dashboard: DashboardStart;
}

export interface ObservabilitySetup {}

export interface ObservabilityStart {}
