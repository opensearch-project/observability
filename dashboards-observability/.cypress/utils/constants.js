/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export const delay = 1500;
export const COMMAND_TIMEOUT_LONG = 10000;

// trace analytics
export const TRACE_ID = '8832ed6abbb2a83516461960c89af49d';
export const SPAN_ID = 'a673bc074b438374';
export const SERVICE_NAME = 'frontend-client';
export const SERVICE_SPAN_ID = '7df5609a6d104736';

export const testDataSet = [
  {
    mapping_url: 'https://raw.githubusercontent.com/opensearch-project/observability/main/dashboards-observability/.cypress/utils/otel-v1-apm-service-map-mappings.json',
    data_url: 'https://raw.githubusercontent.com/opensearch-project/observability/main/dashboards-observability/.cypress/utils/otel-v1-apm-service-map.json',
    index: 'otel-v1-apm-service-map',
  },
  {
    mapping_url: 'https://raw.githubusercontent.com/opensearch-project/observability/main/dashboards-observability/.cypress/utils/otel-v1-apm-span-000001-mappings.json',
    data_url: 'https://raw.githubusercontent.com/opensearch-project/observability/main/dashboards-observability/.cypress/utils/otel-v1-apm-span-000001.json',
    index: 'otel-v1-apm-span-000001',
  },
  {
    mapping_url: 'https://raw.githubusercontent.com/opensearch-project/observability/main/dashboards-observability/.cypress/utils/otel-v1-apm-span-000001-mappings.json',
    data_url: 'https://raw.githubusercontent.com/opensearch-project/observability/main/dashboards-observability/.cypress/utils/otel-v1-apm-span-000002.json',
    index: 'otel-v1-apm-span-000002',
  },
]

export const setTimeFilter = (setEndTime = false, refresh = true) => {
  const startTime = 'Mar 25, 2021 @ 10:00:00.000';
  const endTime = 'Mar 25, 2021 @ 11:00:00.000';
  cy.get('button.euiButtonEmpty[aria-label="Date quick select"]').click();
  cy.get('.euiQuickSelect__applyButton').click();
  cy.get('.euiSuperDatePicker__prettyFormatLink').click();
  cy.get(
    'button.euiDatePopoverButton--start[data-test-subj="superDatePickerstartDatePopoverButton"]'
  ).click();
  cy.get('.euiTab__content').contains('Absolute').click();
  cy.get('input[data-test-subj="superDatePickerAbsoluteDateInput"]')
    .focus()
    .type('{selectall}' + startTime, { force: true });
  if (setEndTime) {
    cy.wait(delay);
    cy.get(
      'button.euiDatePopoverButton--end[data-test-subj="superDatePickerendDatePopoverButton"]'
    ).click();
    cy.wait(delay);
    cy.get('.euiTab__content').contains('Absolute').click();
    cy.get('input[data-test-subj="superDatePickerAbsoluteDateInput"]')
      .focus()
      .type('{selectall}' + endTime, { force: true });
  }
  if (refresh) cy.get('.euiButton__text').contains('Refresh').click();
  cy.wait(delay);
};

// notebooks
export const TEST_NOTEBOOK = 'Test Notebook';
export const SAMPLE_URL = 'https://github.com/opensearch-project/sql/tree/main/sql-jdbc';
export const NOTEBOOK_TEXT = 'Use Notebooks to interactively and collaboratively develop rich reports backed by live data. Common use cases for notebooks includes creating postmortem reports, designing run books, building live infrastructure reports, or even documentation.';
export const OPENSEARCH_URL = 'https://opensearch.org/docs/latest/observability-plugin/notebooks/';
export const MARKDOWN_TEXT = `%md
# Heading 1

#### List and links

* 1
* 2
* [SQL JDBC](${SAMPLE_URL})

---
#### Code block
* Explain SQL
\`\`\`
POST _plugins/_sql/_explain
{
  "query": "SELECT * FROM my-index LIMIT 50"
}
\`\`\`

#### Table
| a1 | b1 | c1 | d1 |
|----|----|----|----|
| a2 | b2 | c2 | d2 |
| a3 | b3 | c3 | d3 |
`

export const SQL_QUERY_TEXT = `%sql
select * from opensearch_dashboards_sample_data_flights limit 20
`

export const PPL_QUERY_TEXT = `%ppl
source=opensearch_dashboards_sample_data_flights
`

export const supressResizeObserverIssue = () => {
  // exception is thrown on loading EuiDataGrid in cypress only, ignore for now
  cy.on('uncaught:exception', (err, runnable) => {
    if (err.message.includes('ResizeObserver loop')) return false;
  });
};

export const verify_traces_spans_data_grid_cols_exists = () => {
  cy.get('.euiDataGridHeaderCell__content').contains('Span ID').should('exist');
  cy.get('.euiDataGridHeaderCell__content').contains('Trace ID').should('exist');
  cy.get('.euiDataGridHeaderCell__content').contains('Operation').should('exist');
  cy.get('.euiDataGridHeaderCell__content').contains('Duration').should('exist');
  cy.get('.euiDataGridHeaderCell__content').contains('Start time').should('exist');
  cy.get('.euiDataGridHeaderCell__content').contains('End time').should('exist');
  cy.get('.euiDataGridHeaderCell__content').contains('Errors').should('exist');
}

export const count_table_row = (expected_row_count) => {
  cy.get('.euiDataGridHeader [role="columnheader"]').then($el => {
    let colmun_header_count = Cypress.$($el).length;
    let table_grid_cell_count = Cypress.$('[data-test-subj="dataGridRowCell"]').length;
    const total_row_count = table_grid_cell_count / colmun_header_count;
    expect(total_row_count).to.equal(expected_row_count)
  });
}
