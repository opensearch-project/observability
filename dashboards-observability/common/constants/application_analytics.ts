/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export const TAB_OVERVIEW_ID = 'app-analytics-overview';
export const TAB_SERVICE_ID = 'app-analytics-service';
export const TAB_TRACE_ID = 'app-analytics-trace';
export const TAB_LOG_ID = 'app-analytics-log';
export const TAB_PANEL_ID = 'app-analytics-panel';
export const TAB_CONFIG_ID = 'app-analytics-config';
export const TAB_OVERVIEW_TITLE = 'Overview';
export const TAB_SERVICE_TITLE = 'Services';
export const TAB_TRACE_TITLE = 'Traces & Spans';
export const TAB_LOG_TITLE = 'Log Events';
export const TAB_PANEL_TITLE = 'Panel';
export const TAB_CONFIG_TITLE = 'Configuration';

export const APP_ANALYTICS_API_PREFIX = '/api/observability/application';
