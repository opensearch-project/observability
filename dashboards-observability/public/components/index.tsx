/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import ReactDOM from 'react-dom';
import { QueryManager } from 'common/query_manager';
import { AppMountParameters, CoreStart } from '../../../../src/core/public';
import { AppPluginStartDependencies } from '../types';
import { App } from './app';

export const Observability = (
  CoreStartProp: CoreStart,
  DepsStart: AppPluginStartDependencies,
  AppMountParametersProp: AppMountParameters,
  pplService: any,
  dslService: any,
  savedObjects: any,
  timestampUtils: any,
  queryManager: QueryManager
) => {
  ReactDOM.render(
    <App
      CoreStartProp={CoreStartProp}
      DepsStart={DepsStart}
      pplService={pplService}
      dslService={dslService}
      savedObjects={savedObjects}
      timestampUtils={timestampUtils}
      queryManager={queryManager}
    />,
    AppMountParametersProp.element
  );

  return () => ReactDOM.unmountComponentAtNode(AppMountParametersProp.element);
};
