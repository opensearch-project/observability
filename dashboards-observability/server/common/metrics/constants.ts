/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { ComponentType, CounterType } from './types';

export const WINDOW = 3600;
export const INTERVAL = 60;
export const CAPACITY = (WINDOW / INTERVAL) * 2;
export const MILLIS_MULTIPLIER = 1000;

export const COMPONENTS = [
  'application_analytics',
  'operational_panels',
  'event_analytics',
  'notebooks',
  'trace_analytics',
  'metrics_analytics',
] as const;
export const REQUESTS = ['create', 'get', 'update', 'delete'] as const;

export const GLOBAL_BASIC_COUNTER: CounterType = (() => {
  const counter = {} as CounterType;
  COMPONENTS.forEach((component) => {
    counter[component] = {} as CounterType[ComponentType];
    REQUESTS.forEach((request) => {
      counter[component][request] = {
        total: 0,
      };
    });
  });
  return counter;
})();

export const DEFAULT_ROLLING_COUNTER: CounterType = (() => {
  const counter = {} as CounterType;
  COMPONENTS.forEach((component) => {
    counter[component] = {} as CounterType[ComponentType];
    REQUESTS.forEach((request) => {
      counter[component][request] = {
        count: 0,
        system_error: 0,
        user_error: 0,
      };
    });
  });
  return counter;
})();
