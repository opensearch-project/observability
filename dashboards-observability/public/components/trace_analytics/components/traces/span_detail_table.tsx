/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 *   Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { EuiDataGrid, EuiDataGridColumn, EuiLink, EuiText } from '@elastic/eui';
import _ from 'lodash';
import moment from 'moment';
import React, { useCallback, useEffect, useMemo, useState } from 'react';
import { HttpSetup } from '../../../../../../../src/core/public';
import { TRACE_ANALYTICS_DATE_FORMAT } from '../../../../../common/constants/trace_analytics';
import { handleSpansRequest } from '../../requests/traces_request_handler';
import { nanoToMilliSec, NoMatchMessage } from '../common/helper_functions';

interface SpanDetailTableProps {
  http: HttpSetup;
  hiddenColumns: string[];
  openFlyout: (spanId: string) => void;
  DSL?: any;
  setTotal?: (total: number) => void;
}

export interface SpanSearchParams {
  from: number;
  size: number;
  sortingColumns: Array<{
    [id: string]: 'asc' | 'desc';
  }>;
}

export function SpanDetailTable(props: SpanDetailTableProps) {
  const [tableParams, setTableParams] = useState({
    size: 10,
    page: 0,
    sortingColumns: [] as Array<{
      id: string;
      direction: 'asc' | 'desc';
    }>,
  });
  const [items, setItems] = useState<any>([]);
  const [total, setTotal] = useState(0);

  useEffect(() => {
    const spanSearchParams: SpanSearchParams = {
      from: tableParams.page * tableParams.size,
      size: tableParams.size,
      sortingColumns: tableParams.sortingColumns.map(({ id, direction }) => ({ [id]: direction })),
    };
    handleSpansRequest(props.http, setItems, setTotal, spanSearchParams, props.DSL);
  }, [tableParams, props.DSL]);

  useEffect(() => {
    if (props.setTotal) props.setTotal(total);
  }, [total]);

  const columns: EuiDataGridColumn[] = [
    {
      id: 'spanId',
      display: 'Span ID',
    },
    {
      id: 'parentSpanId',
      display: 'Parent span ID',
    },
    {
      id: 'traceId',
      display: 'Trace ID',
    },
    {
      id: 'traceGroup',
      display: 'Trace group',
    },
    {
      id: 'serviceName',
      display: 'Service',
    },
    {
      id: 'name',
      display: 'Operation',
    },
    {
      id: 'durationInNanos',
      display: 'Duration',
    },
    {
      id: 'startTime',
      display: 'Start time',
    },
    {
      id: 'endTime',
      display: 'End time',
    },
    {
      id: 'status.code',
      display: 'Errors',
    },
  ];

  const [visibleColumns, setVisibleColumns] = useState(() =>
    columns
      .filter(({ id }) => props.hiddenColumns.findIndex((column) => column === id) === -1)
      .map(({ id }) => id)
  );

  const renderCellValue = useMemo(() => {
    return ({ rowIndex, columnId }) => {
      let adjustedRowIndex = rowIndex - tableParams.page * tableParams.size;
      if (!items.hasOwnProperty(adjustedRowIndex)) return '-';
      const value = items[adjustedRowIndex][columnId];
      if (value == null || value === '') return '-';
      switch (columnId) {
        case 'spanId':
          return <EuiLink onClick={() => props.openFlyout(value)}>{value}</EuiLink>;
        case 'durationInNanos':
          return `${_.round(nanoToMilliSec(Math.max(0, value)), 2)} ms`;
        case 'startTime':
        case 'endTime':
          return moment(value).format(TRACE_ANALYTICS_DATE_FORMAT);
        case 'status.code':
          return value === 2 ? (
            <EuiText color="danger" size="s">
              Yes
            </EuiText>
          ) : (
            'No'
          );

        default:
          return value;
      }
    };
  }, [items, tableParams.page, tableParams.size]);

  const onSort = useCallback(
    (sortingColumns) => {
      setTableParams({
        ...tableParams,
        sortingColumns,
      });
    },
    [setTableParams]
  );

  const onChangeItemsPerPage = useCallback((size) => setTableParams({ ...tableParams, size }), [
    tableParams, setTableParams,
  ]);
  const onChangePage = useCallback((page) => setTableParams({ ...tableParams, page }), [
    tableParams, setTableParams,
  ]);

  return (
    <>
      <EuiDataGrid
        aria-labelledby="span-detail-data-grid"
        columns={columns}
        columnVisibility={{ visibleColumns, setVisibleColumns }}
        rowCount={total}
        renderCellValue={renderCellValue}
        sorting={{ columns: tableParams.sortingColumns, onSort }}
        pagination={{
          pageIndex: tableParams.page,
          pageSize: tableParams.size,
          pageSizeOptions: [10, 50, 100],
          onChangeItemsPerPage: onChangeItemsPerPage,
          onChangePage: onChangePage,
        }}
      />
      {total === 0 && <NoMatchMessage size="xl" />}
    </>
  );
}
