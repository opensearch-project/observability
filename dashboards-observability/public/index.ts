/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import { PluginInitializer, PluginInitializerContext } from '../../../src/core/public';
import {
  ObservabilityPlugin
} from './plugin';

export { ObservabilityPlugin as Plugin };

export const plugin = (initializerContext: PluginInitializerContext) => new ObservabilityPlugin(initializerContext);