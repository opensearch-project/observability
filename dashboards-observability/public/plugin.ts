/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { AppMountParameters, CoreSetup, CoreStart, Plugin } from '../../../src/core/public';
import {
  observabilityID,
  observabilityPluginOrder,
  observabilityTitle,
} from '../common/constants/shared';
import PPLService from './services/requests/ppl';
import DSLService from './services/requests/dsl';
import TimestampUtils from './services/timestamp/timestamp';
import SavedObjects from './services/saved_objects/event_analytics/saved_objects';
import { AppPluginStartDependencies, ObservabilitySetup, ObservabilityStart } from './types';
import { convertLegacyNotebooksUrl } from './components/notebooks/components/helpers/legacy_route_helpers';
import { convertLegacyTraceAnalyticsUrl } from './components/trace_analytics/components/common/legacy_route_helpers';
import { uiSettingsService } from '../common/utils';
import { QueryManager } from '../common/query_manager';
export class ObservabilityPlugin implements Plugin<ObservabilitySetup, ObservabilityStart> {
  public setup(core: CoreSetup): ObservabilitySetup {
    uiSettingsService.init(core.uiSettings, core.notifications);

    // redirect legacy notebooks URL to current URL under observability
    if (window.location.pathname.includes('notebooks-dashboards')) {
      window.location.assign(convertLegacyNotebooksUrl(window.location));
    }

    // redirect legacy trace analytics URL to current URL under observability
    if (window.location.pathname.includes('trace-analytics-dashboards')) {
      window.location.assign(convertLegacyTraceAnalyticsUrl(window.location));
    }

    core.application.register({
      id: observabilityID,
      title: observabilityTitle,
      category: {
        id: 'opensearch',
        label: 'OpenSearch Plugins',
        order: 2000,
      },
      order: observabilityPluginOrder,
      async mount(params: AppMountParameters) {
        const { Observability } = await import('./components/index');
        const [coreStart, depsStart] = await core.getStartServices();
        const pplService = new PPLService(coreStart.http);
        const dslService = new DSLService(coreStart.http);
        const savedObjects = new SavedObjects(coreStart.http);
        const timestampUtils = new TimestampUtils(dslService, pplService);
        const qm = new QueryManager();
        return Observability(
          coreStart,
          depsStart as AppPluginStartDependencies,
          params,
          pplService,
          dslService,
          savedObjects,
          timestampUtils,
          qm
        );
      },
    });

    // Return methods that should be available to other plugins
    return {};
  }
  public start(core: CoreStart): ObservabilityStart {
    return {};
  }
  public stop() {}
}
