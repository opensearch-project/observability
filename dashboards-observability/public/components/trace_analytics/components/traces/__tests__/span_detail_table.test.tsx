/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { waitFor } from '@testing-library/react';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import ReactDOM from 'react-dom';
import { act } from 'react-dom/test-utils';
import { HttpResponse } from '../../../../../../../../src/core/public';
import { TEST_SPAN_RESPONSE } from '../../../../../../test/constants';
import httpClientMock from '../../../../../../test/__mocks__/httpClientMock';
import { SpanDetailTable } from '../span_detail_table';

describe('<SpanDetailTable /> spec', () => {
  configure({ adapter: new Adapter() });

  it('renders the empty component', async () => {
    httpClientMock.post = jest.fn(() =>
      Promise.resolve(({ hits: { hits: [], total: { value: 0 } } } as unknown) as HttpResponse)
    );
    const utils = await mount(
      <SpanDetailTable
        http={httpClientMock}
        hiddenColumns={['traceId', 'traceGroup']}
        DSL={{}}
        openFlyout={() => {}}
      />
    );
    utils.update();
    await waitFor(() => {
      expect(utils).toMatchSnapshot();
    });
  });

  it('renders the component with data', async () => {
    const setCurrentSpan = jest.fn();
    httpClientMock.post = jest.fn(() =>
      Promise.resolve((TEST_SPAN_RESPONSE as unknown) as HttpResponse)
    );
    let container = document.createElement('div');
    await act(() => {
      ReactDOM.render(
        <SpanDetailTable
          http={httpClientMock}
          hiddenColumns={['traceId', 'traceGroup']}
          DSL={{}}
          openFlyout={(spanId: string) => setCurrentSpan(spanId)}
        />,
        container
      );
    });
    expect(container).toMatchSnapshot();
  });
});
