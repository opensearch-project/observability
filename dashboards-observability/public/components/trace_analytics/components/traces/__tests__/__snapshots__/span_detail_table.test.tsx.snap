// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SpanDetailTable /> spec renders the component with data 1`] = `
<div>
  <div
    data-focus-guard="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="-1"
  />
  <div
    data-focus-guard="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="-1"
  />
  <div
    class="euiDataGrid__focusWrap"
    data-focus-lock-disabled="disabled"
  >
    <div
      class="euiDataGrid euiDataGrid--bordersAll euiDataGrid--headerShade euiDataGrid--footerOverline euiDataGrid--rowHoverHighlight euiDataGrid--stickyFooter"
    />
  </div>
  <div
    data-focus-guard="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="-1"
  />
</div>
`;

exports[`<SpanDetailTable /> spec renders the empty component 1`] = `
<SpanDetailTable
  DSL={Object {}}
  hiddenColumns={
    Array [
      "traceId",
      "traceGroup",
    ]
  }
  http={[MockFunction]}
  openFlyout={[Function]}
>
  <EuiDataGrid
    aria-labelledby="span-detail-data-grid"
    columnVisibility={
      Object {
        "setVisibleColumns": [Function],
        "visibleColumns": Array [
          "spanId",
          "parentSpanId",
          "serviceName",
          "name",
          "durationInNanos",
          "startTime",
          "endTime",
          "status.code",
        ],
      }
    }
    columns={
      Array [
        Object {
          "display": "Span ID",
          "id": "spanId",
        },
        Object {
          "display": "Parent span ID",
          "id": "parentSpanId",
        },
        Object {
          "display": "Trace ID",
          "id": "traceId",
        },
        Object {
          "display": "Trace group",
          "id": "traceGroup",
        },
        Object {
          "display": "Service",
          "id": "serviceName",
        },
        Object {
          "display": "Operation",
          "id": "name",
        },
        Object {
          "display": "Duration",
          "id": "durationInNanos",
        },
        Object {
          "display": "Start time",
          "id": "startTime",
        },
        Object {
          "display": "End time",
          "id": "endTime",
        },
        Object {
          "display": "Errors",
          "id": "status.code",
        },
      ]
    }
    pagination={
      Object {
        "onChangeItemsPerPage": [Function],
        "onChangePage": [Function],
        "pageIndex": 0,
        "pageSize": 10,
        "pageSizeOptions": Array [
          10,
          50,
          100,
        ],
      }
    }
    renderCellValue={[Function]}
    rowCount={0}
    sorting={
      Object {
        "columns": Array [],
        "onSort": [Function],
      }
    }
  >
    <EuiI18n
      default="{label}; Page {page} of {pageCount}."
      token="euiDataGrid.ariaLabel"
      values={
        Object {
          "label": undefined,
          "page": 1,
          "pageCount": 0,
        }
      }
    >
      <EuiI18n
        default="Page {page} of {pageCount}."
        token="euiDataGrid.ariaLabelledBy"
        values={
          Object {
            "page": 1,
            "pageCount": 0,
          }
        }
      >
        <EuiFocusTrap
          className="euiDataGrid__focusWrap"
          disabled={true}
        >
          <ForwardRef
            className="euiDataGrid__focusWrap"
            enabled={false}
            noIsolation={true}
            returnFocus={true}
            scrollLock={false}
          >
            <ForwardRef
              className="euiDataGrid__focusWrap"
              enabled={false}
              noIsolation={true}
              returnFocus={true}
              scrollLock={false}
              sideCar={[Function]}
            >
              <ForwardRef
                as={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "classNames": Object {
                      "fullWidth": "width-before-scroll-bar",
                      "zeroRight": "right-scroll-bar-position",
                    },
                    "defaultProps": Object {
                      "enabled": true,
                      "inert": false,
                      "removeScrollBar": true,
                    },
                    "render": [Function],
                  }
                }
                autoFocus={true}
                className="euiDataGrid__focusWrap"
                crossFrame={true}
                disabled={true}
                lockProps={
                  Object {
                    "allowPinchZoom": undefined,
                    "as": undefined,
                    "enabled": false,
                    "inert": undefined,
                    "shards": undefined,
                    "sideCar": [Function],
                    "style": undefined,
                  }
                }
                noFocusGuards={false}
                persistentFocus={false}
                returnFocus={true}
                sideCar={[Function]}
              >
                <div
                  data-focus-guard={true}
                  key="guard-first"
                  style={
                    Object {
                      "height": "0px",
                      "left": "1px",
                      "overflow": "hidden",
                      "padding": 0,
                      "position": "fixed",
                      "top": "1px",
                      "width": "1px",
                    }
                  }
                  tabIndex={-1}
                />
                <div
                  data-focus-guard={true}
                  key="guard-nearest"
                  style={
                    Object {
                      "height": "0px",
                      "left": "1px",
                      "overflow": "hidden",
                      "padding": 0,
                      "position": "fixed",
                      "top": "1px",
                      "width": "1px",
                    }
                  }
                  tabIndex={-1}
                />
                <ForwardRef
                  className="euiDataGrid__focusWrap"
                  data-focus-lock-disabled="disabled"
                  enabled={false}
                  inert={false}
                  onBlur={[Function]}
                  onFocus={[Function]}
                  removeScrollBar={true}
                  sideCar={[Function]}
                >
                  <div
                    className="euiDataGrid__focusWrap"
                    data-focus-lock-disabled="disabled"
                    onBlur={[Function]}
                    onFocus={[Function]}
                    onScrollCapture={[Function]}
                    onTouchMoveCapture={[Function]}
                    onWheelCapture={[Function]}
                  >
                    <div
                      className="euiDataGrid euiDataGrid--bordersAll euiDataGrid--headerShade euiDataGrid--footerOverline euiDataGrid--rowHoverHighlight euiDataGrid--stickyFooter"
                      onKeyDown={[Function]}
                    />
                  </div>
                </ForwardRef>
                <div
                  data-focus-guard={true}
                  style={
                    Object {
                      "height": "0px",
                      "left": "1px",
                      "overflow": "hidden",
                      "padding": 0,
                      "position": "fixed",
                      "top": "1px",
                      "width": "1px",
                    }
                  }
                  tabIndex={-1}
                />
              </ForwardRef>
            </ForwardRef>
          </ForwardRef>
        </EuiFocusTrap>
      </EuiI18n>
    </EuiI18n>
  </EuiDataGrid>
  <NoMatchMessage
    size="xl"
  >
    <EuiSpacer
      size="xl"
    >
      <div
        className="euiSpacer euiSpacer--xl"
      />
    </EuiSpacer>
    <EuiEmptyPrompt
      body={
        <EuiText>
          No data matches the selected filter. Clear the filter and/or increase the time range to see more results.
        </EuiText>
      }
      title={
        <h2>
          No matches
        </h2>
      }
    >
      <div
        className="euiEmptyPrompt"
      >
        <EuiTextColor
          color="subdued"
        >
          <span
            className="euiTextColor euiTextColor--subdued"
          >
            <EuiTitle
              size="m"
            >
              <h2
                className="euiTitle euiTitle--medium"
              >
                No matches
              </h2>
            </EuiTitle>
            <EuiSpacer
              size="m"
            >
              <div
                className="euiSpacer euiSpacer--m"
              />
            </EuiSpacer>
            <EuiText>
              <div
                className="euiText euiText--medium"
              >
                <EuiText>
                  <div
                    className="euiText euiText--medium"
                  >
                    No data matches the selected filter. Clear the filter and/or increase the time range to see more results.
                  </div>
                </EuiText>
              </div>
            </EuiText>
          </span>
        </EuiTextColor>
      </div>
    </EuiEmptyPrompt>
    <EuiSpacer
      size="xl"
    >
      <div
        className="euiSpacer euiSpacer--xl"
      />
    </EuiSpacer>
  </NoMatchMessage>
</SpanDetailTable>
`;
